signUrl/SignUrlSample/Program.cs

```diff
@@ -0,0 +1,119 @@
+        public string Url { get; set; }
+
+        [Option("name", HelpText = "The name of the key to use when encrypting url")]
+        public string KeyName { get; set; }
+
+        [Option("path", HelpText = "The path to the key to use when encrypting url" )]
+        public string KeyPath { get; set; }
+
+        [Option('e', HelpText = "Url expiration in UTC formatted as: YYYY-MM-DDThh:mm:ss")]
+        public DateTime Expiration { get; set; }
+    }
+
+    public class Program
+    {
+        static int Main(string[] args)
+        {
+            string signedUrl = null;
+            Parser.Default.ParseArguments<SignedUrlArgs>(args).MapResult(
+                uargs => signedUrl = CreateSignedUrl(uargs.Url, uargs.KeyName, File.ReadAllText(uargs.KeyPath, Encoding.UTF8), uargs.Expiration), 
+                err => string.Empty);
+
+            if (string.IsNullOrEmpty(signedUrl))
+            {
+                return 1;
+            }
+
+            Console.WriteLine($"Signed URL: {signedUrl}");
+            return 0;
+        }
+
+        // [START example]
+        /// <summary>
+        /// Creates signed URL for Google Cloud SDN
+        /// More details about order of operations is here: <see cref="https://cloud-dot-devsite.googleplex.com/cdn/docs/signed-urls#creatingkeys"/>
+        /// </summary>
+        /// <param name="url">The Url to sign. This URL can't include Expires and KeyName query parameters in it</param>
+        /// <param name="keyName">The name of the key used to sign the URL</param>
+        /// <param name="encodedKey">The key used to sign the Url</param>
+        /// <param name="expirationTime">Expiration time of the signature</param>
+        /// <returns>Signed Url that is valid until {expirationTime}</returns>
+        public static string CreateSignedUrl(string url, string keyName, string encodedKey, DateTime expirationTime)
+        {
+            var builder = new UriBuilder(url);
+
+            long unixTimestampExpiration = ToUnixTime(expirationTime);
+
+            char queryParam = string.IsNullOrEmpty(builder.Query) ? '?' : '&';
+            builder.Query += $"{queryParam}Expires={unixTimestampExpiration}&KeyName={keyName}".ToString();
+
+            // Key is passed as base64url encoded
+            byte[] decodedKey = Base64UrlDecode(encodedKey);
+
+            // Computes SHA-1 hash of the URL using the key
+            byte[] hash = ComputeHash(decodedKey, builder.Uri.AbsoluteUri);
+            string encodedHash = Base64UrlEncode(hash);
+            
+            builder.Query += $"&Signature={encodedHash}";
+            return builder.Uri.AbsoluteUri;
+        }
+
+        private static long ToUnixTime(DateTime date)
+        {
+            var epoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
+            return Convert.ToInt64((date - epoch).TotalSeconds);
+        }
+
+        private static byte[] Base64UrlDecode(string arg)
+        {
+            string s = arg;
```

Comment:
new variable declaration is not necessary

Reply:
If I remove this assignment, I'll have to change input parameter value in the method, which is bad practice if you don't explicitly declare this as intent by making the method signature require passing parameter by reference:
<b>Base64UrlDecode(ref string arg);</b>

In this case we return byte array, so no point in changing method signature.
Even though strings are immutable in C# and have value type semantics, still it's better not to mutate input parameter.

Reply:
Maybe I'm recalling this wrong about C#, but here's an experiment: https://dotnetfiddle.net/o9EV5F

When you pass a string to a function, and modify its value, its value in the outer scope remains the same.